### PR TITLE
Fix memory bugs in loading code

### PR DESCRIPTION
**Pull request by jart**
This change hardens the C++ code that loads the GGML file format. Some people download weights off the Internet to run inference on a trained model. Since weights don't contain code like graph definitions, having them be able to load in a secure manner is a reasonable expectation to have. Therefore this change addresses many of the weaknesses in how we were going about doing things earlier, which would allow untrustworthy weights to trigger undefined behaviors with memory. I haven't cared to investigate whether any of these weaknesses are exploitable, but it'll certainly be more difficult for that to happen, once this gets merged, which will enable our users to share more freely, happily, and safely.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -316,17 +339,26 @@ static void *mmap_file(const char *fname, uint64_t *mm_length) {
                                OPEN_EXISTING,
                                FILE_ATTRIBUTE_NORMAL | FILE_ATTRIBUTE_NOT_CONTENT_INDEXED,
                                NULL);
-    if (hFile == INVALID_HANDLE_VALUE) return 0;
+    if (hFile == INVALID_HANDLE_VALUE) {
+        LogWindowsError("CreateFileA");
+        return 0;
```
Comment (howard0su) [inline llama.cpp:344]: return NULL as void* is a ptr.

File: llama.cpp [hunk 2]
```diff
@@ -359,6 +391,46 @@ static bool report_bad_magic(const char *path, uint32_t got, uint32_t want) {
...
+        fprintf(stderr, "%s failed: system error: %s\n", thing, strerror(errno));
+    } else if (ios.eof()) {
+        fprintf(stderr, "%s failed: unexpected end of file\n", thing);
+    } else if (ios.fail()) {
+        fprintf(stderr, "%s failed: illogical operation\n", thing);
+    }
+    return false;
+}
+
+static bool read_impl(std::istream &fin, char *buf, std::streamsize len, const char *thing) {
```
Comment (howard0su) [inline llama.cpp:417]: I don't understand there. if you already validate the size of file, how could read int/float/buf can fail? fail by what?

File: llama.cpp [hunk 3]
```diff
@@ -476,20 +546,20 @@ static bool llama_model_load(
...
-                fin.read(tmp.data(), len);
+                if (!read_buf(fin, tmp.data(), len)) return false;
                 word.assign(tmp.data(), len);
             } else {
                 word.clear();
             }
 
             float score;
-            fin.read((char *) &score, sizeof(score));
+            if (!read_float(fin, &score)) return false;
```
Comment (howard0su) [inline llama.cpp:562]: check nan here. I don't think read will fail.

File: llama.cpp [hunk 4]
```diff
@@ -385,13 +457,13 @@ static bool llama_model_load(
     fin.rdbuf()->pubsetbuf(f_buf.data(), f_buf.size());
 
     fin.seekg(0, fin.end);
-    const size_t file_size = fin.tellg();
+    const int64_t file_size = fin.tellg();
```
Comment (howard0su) [inline llama.cpp:460]: check the file_size. we should have a very precise file size for different models.

File: llama.cpp [hunk 5]
```diff
@@ -546,15 +615,15 @@ static bool llama_model_load(
         const size_t scale = memory_type == GGML_TYPE_F32 ? 2 : 1;
 
         // this is the total memory required to run the inference
-        const size_t mem_required =
+        const int64_t mem_required =
```
Comment (howard0su) [inline llama.cpp:618]: why? size_t is correct here as we are calculate the memory size.

File: llama.cpp [hunk 6]
```diff
@@ -641,31 +710,43 @@ static bool llama_model_load(
...
 
-            fin.read(reinterpret_cast<char *>(&n_dims), sizeof(n_dims));
-            fin.read(reinterpret_cast<char *>(&length), sizeof(length));
-            fin.read(reinterpret_cast<char *>(&ftype),  sizeof(ftype));
-
-            if (fin.eof()) {
-                break;
-            }
+            fin.read((char *)&n_dims, 4);
+            if (fin.eof()) break;
```
Comment (howard0su) [inline llama.cpp:722]: as I mentioned earlier, please check the size intially. check eof is not good idea for every io.

File: llama.cpp [hunk 7]
```diff
@@ -641,31 +710,43 @@ static bool llama_model_load(
...
+            if (!fin.good()) return false;
+            if (!read_int32(fin, &length)) return false;
+            if (!read_int32(fin, &ftype)) return false;
 
             int32_t nelements = 1;
             int32_t ne[2] = { 1, 1 };
+            if (!check_n_dims(n_dims)) return false;
```
Comment (howard0su) [inline llama.cpp:729]: this check is good.

File: llama.cpp [hunk 8]
```diff
@@ -1508,18 +1576,14 @@ static bool llama_model_quantize_internal(const std::string & fname_inp, const s
 
                 switch (type) {
                     case GGML_TYPE_Q4_0:
-                        {
-                            cur_size = ggml_quantize_q4_0(data_f32.data(), work.data(), nelements, ne[0], hist_cur.data());
-                        } break;
+                        cur_size = ggml_quantize_q4_0(data_f32.data(), work.data(), nelements, ne[0], hist_cur.data());
```
Comment (howard0su) [inline llama.cpp:1579]: not related. better a separate PR

**Review by howard0su — COMMENTED** [review]
Overall, i suggest removing all check around read. do validate the incoming number is in the valid range. This will also make the PR smaller.

File: llama.cpp [hunk 9]
```diff
@@ -1508,18 +1576,14 @@ static bool llama_model_quantize_internal(const std::string & fname_inp, const s
...
-                        {
-                            cur_size = ggml_quantize_q4_1(data_f32.data(), work.data(), nelements, ne[0], hist_cur.data());
-                        } break;
+                        cur_size = ggml_quantize_q4_1(data_f32.data(), work.data(), nelements, ne[0], hist_cur.data());
+                        break;
                     default:
-                        {
-                            fprintf(stderr, "%s: unsupported quantization type %d\n", __func__, type);
-                            return false;
-                        }
```
Comment (ggerganov) [inline llama.cpp:1522]: The `switch` format using `{}` has to remain as it is

**Review by ggerganov — APPROVED** [review]
I consider such kind of changes with very low priority and almost unnecessary.

The reason is that this is only really needed when making a production-ready software where you need to care about wether your files got corrupted or someone could be trying to exploit the software, etc.

This project is not that, since we are not expecting it to be used in serious / commercial products. `llama.cpp` is mostly a playground for exploring inference techniques and applications of LLMs. So we want to keep things simple, even at the cost of some good programming practices and proper input sanitisation.

At a later stage, when and if we start building a proper `ggml`-based engine, we can start considering this kind of safety issues more seriously. But for now, it's not really worth it.

Since you've already put the work into making this, I am approving it.
But unless you feel super strong about adding these changes, I recommend to close the PR

**Pull request closed without merging**